File: src/main/java/pranav/views/Background.java
package pranav.views;

import android.animation.ObjectAnimator;
import android.animation.TimeInterpolator;
import android.content.Context;
import android.support.annotation.FloatRange;
import android.support.annotation.IdRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.view.KeyEvent;
import android.widget.FrameLayout;

import pranav.utilities.ArgbEval;

import static pranav.utilities.Animations.ANIMATION_TIME;
import static pranav.utilities.Utilities.MATCH;
import static pranav.utilities.Utilities.isFinite;

/**
 * Created on 10-07-2017 at 21:58 by Pranav Raut.
 * For QRCodeProtection
 */

public class Background extends FrameLayout  {

    @IdRes
    public static final int BACK_ID = 2146;
    private int color = 0x90000000;
    private float percent;

    private ArgbEval argbEval = new ArgbEval(0, color);
    private ObjectAnimator animator;

    private FrameLayout listener;
    @Nullable
    private Keys keys;
    private TimeInterpolator interpolator;
    private Listeners.l1<Background> l;
    private int state;

    public Background(@NonNull Context context) {
        super(context);
        init();
    }

    public Background(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    private void init() {
        setLayoutParams(MATCH);
        setFocusableInTouchMode(true);
        setFocusable(true);
        listener = new FrameLayout(getContext());
        listener.setLayoutParams(MATCH);
        super.addView(listener);
        argbEval.setObject(this);
        listener.setId(BACK_ID);
        l = new Listeners.l1<>(this);
    }

    public Background setClickListener(@Nullable final OnClickListener l) {
        listener.setOnClickListener(l);
        return this;
    }

    public void setColor(int color) {
        this.color = color;
        argbEval = new ArgbEval(0, color);
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        if (keyCode == KeyEvent.KEYCODE_BACK && event.getAction() == KeyEvent.ACTION_DOWN)
            return onBackPress(keyCode, event);
        return super.onKeyDown(keyCode, event);
    }

    public Background setKeys(@Nullable Keys onBackPress) {
        this.keys = onBackPress;
        return this;
    }

    private boolean onBackPress(int keyCode, KeyEvent event) {
        boolean b = keys == null || keys.onBackPress(percent, keyCode, event);
        clearFocus();
        return b;
    }

    public void animateTo(@FloatRange(from = 0, to = 100) float percent) {
        if (percent != this.percent)
            if (animator == null || !animator.isStarted()) {
                animator = ObjectAnimator.ofFloat(this, "percent", 100 - percent, percent);
                animator.setDuration(ANIMATION_TIME).setInterpolator(interpolator);
                animator.addListener(l.getAnimator());
                animator.start();
            }
    }

    void addListener(L1... l1s) {
        l.addListener(l1s);
    }

    void remove(L1... l1s) {
        l.removeListener(l1s);
    }

    public void setInterpolator(TimeInterpolator interpolator) {
        this.interpolator = interpolator;
    }

    public void toggle() {
        if (!isRunning()) animateTo(100 - percent);
    }

    public float getPercent() {
        return percent;
    }

    public void setPercent(@FloatRange(from = 0f, to = 100f) float percent) {
        if (isFinite(percent)) {
            if (percent == 100) setState(VISIBLE);
            else if (percent == 0) setState(GONE);
            argbEval.getValue(percent / 100f);
            this.percent = percent;
        }
    }

    public int getState() {
        return state;
    }

    public Background setState(@vi int visibility) {
        listener.setVisibility(state = visibility);
        animator = null;
        if (visibility == VISIBLE) {
            percent = 100;
            requestFocus();
        } else percent = 0;
        return this;
    }

    public boolean isRunning() {
        return l.isRunning();
    }

    public Listeners.l1<Background> getL() {
        return l;
    }

}

File: src/main/java/pranav/views/Background.kt
package pranav.views

import android.animation.ObjectAnimator
import android.animation.TimeInterpolator
import android.content.Context
import android.os.Build
import android.util.AttributeSet
import android.view.KeyEvent
import android.view.View
import android.widget.FrameLayout
import androidx.annotation.FloatRange
import androidx.annotation.RequiresApi
import pranav.utilities.Animations.ANIMATION_DURATION
import pranav.utilities.ArgbEval
import pranav.utilities.Utilities.MATCH
import java.lang.Double.isFinite

/**
 * Created on 10-07-2017 at 21:58 by Pranav Raut.
 * For QRCodeProtection
 */

class Background : FrameLayout {
    private var color = -0x70000000
    private var percent: Float = 0.toFloat()

    private var argbEval = ArgbEval(0, color)
    private var animator: ObjectAnimator? = null

    private var clickCaptureView: FrameLayout? = null
    private var keys: BackKeyListener? = null
    private var interpolator: TimeInterpolator? = null
    var l: Listeners.l1<Background>? = null
        private set
    private var state: Int = 0

    val isRunning: Boolean
        get() = l!!.isRunning

    constructor(context: Context) : super(context) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        init()
    }

    private fun init() {
        layoutParams = MATCH
        isFocusableInTouchMode = true
        isFocusable = true
        clickCaptureView = FrameLayout(context)
        clickCaptureView!!.layoutParams = MATCH
        super.addView(clickCaptureView)
        argbEval.setObject(this)
        clickCaptureView!!.id = BACK_ID
        l = Listeners.l1(this)
    }

    fun setClickListener(l: View.OnClickListener?): Background {
        clickCaptureView!!.setOnClickListener(l)
        return this
    }

    fun setColor(color: Int): Background {
        this.color = color
        argbEval = ArgbEval(0, color)
        return this
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        return if (keyCode == KeyEvent.KEYCODE_BACK && event.action == KeyEvent.ACTION_DOWN) onBackPress(
            keyCode,
            event
        ) else super.onKeyDown(
            keyCode,
            event
        )
    }

    fun setKeys(onBackPress: BackKeyListener?): Background {
        this.keys = onBackPress
        return this
    }

    private fun onBackPress(keyCode: Int, event: KeyEvent): Boolean {
        val b = keys == null || keys!!.onBackPress(percent, keyCode, event)
        clearFocus()
        return b
    }

    fun animateTo(@FloatRange(from = 0.0, to = 100.0) percent: Float) {
        if (percent != this.percent)
            if (animator == null || !animator!!.isStarted) {
                animator = ObjectAnimator.ofFloat(this, "percent", 100 - percent, percent)
                animator!!.setDuration(ANIMATION_DURATION).interpolator = interpolator
                animator!!.addListener(l!!.animator)
                animator!!.start()
            }
    }

    internal fun addListener(vararg l1s: L1) {
        l!!.addListener(*l1s)
    }

    internal fun removeListener(vararg l1s: L1) {
        l!!.removeListener(*l1s)
    }

    fun setInterpolator(interpolator: TimeInterpolator) {
        this.interpolator = interpolator
    }

    fun toggle() {
        if (!isRunning) animateTo(100 - percent)
    }

    fun getPercent(): Float {
        return percent
    }

    @RequiresApi(Build.VERSION_CODES.N)
    fun setPercent(@FloatRange(from = 0.0, to = 100.0) percent: Float) {
        if (isFinite(percent.toDouble())) {
            if (percent == 100f)
                setState(View.VISIBLE)
            else if (percent == 0f) setState(View.GONE)
            argbEval.getValue(percent / 100f)
            this.percent = percent
        }
    }

    fun getState(): Int {
        return state
    }

    fun setState(@vi visibility: Int): Background {
        clickCaptureView!!.visibility = visibility
        animator = null
        if (visibility == View.VISIBLE) {
            percent = 100f
            requestFocus()
        } else
            percent = 0f
        return this
    }

    companion object {

        val BACK_ID = 2146
    }
}
